### PR TITLE
Typescript documentation

### PR DESCRIPTION
**Pull request by djhi**

### CODE REVIEW

File: docs/css/prism.css [hunk 1]
```diff
@@ -1,139 +1,118 @@
...
-code[class*="language-"]::selection, code[class*="language-"] ::selection {
-	text-shadow: none;
-	background: #b3d4fc;
-}
-
+  color: #000;
+  background: 0 0;
+  text-shadow: 0 1px #fff;
+  font-family: Consolas, Monaco, "Andale Mono", "Ubuntu Mono", monospace;
+  font-size: 1em;
```
Comment (fzaninotto) [inline docs/css/prism.css:9]: I think this will change the size of the code snippets

File: docs/css/prism.css [hunk 2]
```diff
@@ -1,139 +1,118 @@
...
+  }
 }
-
-/* Code blocks */
 pre[class*="language-"] {
-	padding: 1em;
-	margin: .5em 0 1.275em 0;
-	overflow: auto;
+  padding: 1em;
+  margin: 0.5em 0;
```
Comment (fzaninotto) [inline docs/css/prism.css:46]: This was probably adjusted by hand

File: docs/js/prism.js
Comment (fzaninotto) [inline docs/js/prism.js:1]: You're replacing a minified version with an unminified one

File: docs/useGetList.md [hunk 3]
```diff
@@ -118,3 +118,40 @@ const LatestNews = () => {
...
+    );
+    if (isLoading) { return <Loading />; }
+    if (error) { return <p>ERROR</p>; }
+    return (
+        <>
+            <h1>Latest news</h1>
+            <ul>
+                {data.map(record =>
+                    <li key={record.id}>{record.title}</li>
+                )}
```
Comment (fzaninotto) [inline docs/useGetList.md:151]:
```suggestion
                {data.map(post =>
                    <li key={post.id}>{post.title}</li>
                )}
```

Same for the other snippets

File: docs/useGetList.md [hunk 4]
```diff
@@ -118,3 +118,40 @@ const LatestNews = () => {
 ```
 
 The `data` will automatically update when a new record is created, or an existing record is updated or deleted.
+
+## TypeScript
+
+The `useGetList` hook accepts a generic parameter for the record type:
```
Comment (fzaninotto) [inline docs/useGetList.md:124]: You should explain the benefit that this brings in text before the snippet.

File: docs/useGetMany.md [hunk 5]
```diff
@@ -49,3 +49,41 @@ dataProvider.getMany('tags', { ids: [1, 2, 3, 4, 5, 6, 7] });
...
+type Tag = {
+    id: number;
+    name: string;
+}
+
+const PostTags = () => {
+    // record is of type Post
+    const record = useRecordContext<Post>();
+    // data is of type Tag[]
+    const { data, isLoading, error } = useGetMany(
```
Comment (fzaninotto) [inline docs/useGetMany.md:75]: Missing type here?

File: docs/useGetList.md [hunk 6]
```diff
@@ -118,3 +118,40 @@ const LatestNews = () => {
...
+```tsx
+import { useGetList } from 'react-admin';
+
+type Post = {
+    id: number;
+    title: string;
+};
+
+const LatestNews = () => {
+    // data is of type Post[]
```
Comment (fzaninotto) [inline docs/useGetList.md:135]: I'd put that comment lower in the snippet, when you actually use the data fields. And I'd rephrase it to something like

> TypeScript knows that `data` is of type Post[]

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: docs/css/prism.css (hunk 2 above)
Comment (djhi) [reply docs/css/prism.css:46]: Same

File: docs/js/prism.js
Comment (djhi) [reply docs/js/prism.js:1]: Fixed

File: docs/useGetList.md (hunk 4 above)
Comment (djhi) [reply docs/useGetList.md:124]: Explain the benefit of typescript?

File: docs/useDataProvider.md [hunk 7]
```diff
@@ -48,12 +48,56 @@ But the recommended way to query the Data Provider is to use the dataProvider me
...
-**Tip**: If you use TypeScript, you can specify a record type for more type safety:
+## TypeScript
+
+The `useDataProvider` hook accepts a generic parameter for the `dataProvider` type. This is useful when you added custom methods to your `dataProvider`:
+
+```tsx
+// In src/dataProvider.ts
+import { DataProvider } from 'react-admin';
+
+export interface CustomDataProviderMethods extends DataProvider {
```
Comment (fzaninotto) [inline docs/useDataProvider.md:59]: The interface name is misleading. `AugmentedDataProvider` or `DataProviderWithCustomMethods` would be better.

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Excellent!

File: docs/useGetMany.md [hunk 8]
```diff
@@ -49,3 +49,42 @@ dataProvider.getMany('tags', { ids: [1, 2, 3, 4, 5, 6, 7] });
...
+    id: number;
+    name: string;
+}
+
+const PostTags = () => {
+    const post = useRecordContext<Post>();
+
+    const { data: tags, isLoading, error } = useGetMany<Tag>(
+        'tags',
+        // TypeScript knows that post is of type Post
```
Comment (fzaninotto) [inline docs/useGetMany.md:76]: I find this comment disturbing. It's drawing my attention to something else than what I'm supposed to learn here. I'd remove it. Same for other locations where you use `useRecordContext` with a type.

Comment (djhi) [inline docs/useGetMany.md:76]:
```suggestion
```

File: docs/useGetManyReference.md [hunk 9]
```diff
@@ -84,3 +84,48 @@ const PostComments = () => {
...
+}
+
+const PostComments = () => {
+    const post = useRecordContext<Post>();
+    // fetch all comments related to the current record
+    const { data: comments, isLoading, error } = useGetManyReference<Comment>(
+        'comments',
+        { 
+            target: 'post_id',
+            // TypeScript knows that post is of type Post
```
Comment (djhi) [inline docs/useGetManyReference.md:114]:
```suggestion
```

File: docs/useGetOne.md [hunk 10]
```diff
@@ -128,3 +128,32 @@ const UserProfile = () => {
...
+};
+
+type User = {
+    id: number;
+    username: string;
+}
+
+const UserProfile = () => {
+    const ticket = useRecordContext<Ticket>();
+    // TypeScript knows that ticket is of type Ticket
```
Comment (djhi) [inline docs/useGetOne.md:152]:
```suggestion
```

File: docs/useGetOneLive.md [hunk 11]
```diff
@@ -29,3 +29,33 @@ const UserProfile = () => {
...
+};
+
+type User = {
+    id: number;
+    username: string;
+}
+
+const UserProfile = () => {
+    const ticket = useRecordContext<Ticket>();
+    // TypeScript knows that ticket is of type Ticket
```
Comment (djhi) [inline docs/useGetOneLive.md:54]:
```suggestion
```
